recirq/qcqmc/trial_wf.py

```diff
@@ -0,0 +1,1584 @@
+        if np.isclose(abs(w_full[i] - w_full[j]), 0):
+            eigs_diff[i, j] = 1
+        else:
+            eigs_diff[i, j] = (np.exp(1j * (w_full[i] - w_full[j])) - 1) / (
+                1j * (w_full[i] - w_full[j])
+            )
+
+    Y_full = np.zeros_like(v_full, dtype=np.complex128)
+    if a == b:
+        Y_full[a, b] = 0
+    else:
+        Y_full[a, b] = 1.0
+        Y_full[b, a] = -1.0
+
+    Y_kl_full = v_full.conj().T @ Y_full @ v_full
+    # now rotate Y_{kl} * (exp(i(l_{k} - l_{l})) - 1) / (i(l_{k} - l_{l}))
+    # into the original basis
+    pre_matrix_full = v_full @ (eigs_diff * Y_kl_full) @ v_full.conj().T
+
+    return pre_matrix_full
+
+
+def evaluate_gradient_and_cost_function(
+    initial_wf: FqeWavefunction,
+    fqe_ham: RestrictedHamiltonian,
+    n_orb: int,
+    one_body_params: npt.NDArray[np.float64],
+    two_body_params: npt.NDArray[np.float64],
+    gate_generators: List[of.FermionOperator],
+    restricted: bool,
+    e_core: float,
+) -> Tuple[float, npt.NDArray[np.float64]]:
+    """Evaluate gradient and cost function for optimization.
+
+    Args:
+        initial_wf: Initial state (typically Hartree--Fock).
+        fqe_ham: The restricted Hamiltonian in FQE format.
+        n_orb: The number of spatial orbitals.
+        one_body_params: The parameters of the single-particle rotations.
+        two_body_params: The parameters for the two-particle terms.
+        gate_generators: The generators for the two-particle terms.
+        retricted: Whether the single-particle rotations are restricted (True)
+            or unrestricted (False). Unrestricted implies different parameters
+            for the alpha- and beta-spin rotations.
+    Returns:
+        cost_val: The cost function (total energy) evaluated for the input wavefunction parameters.
+        grad: An array of gradients with respect to the one- and two-body
+            parameters. The first n_orb * (n_orb + 1) // 2 parameters correspond to
+            the one-body gradients.
+    """
+    phi = get_evolved_wf(
+        one_body_params,
+        two_body_params,
+        initial_wf,
+        gate_generators,
+        n_orb,
+        restricted=restricted,
+    )[0]
+    lam = copy.deepcopy(phi)
+    lam = lam.apply(fqe_ham)
+    cost_val = fqe.vdot(lam, phi) + e_core
+
+    # 1body
+    one_body_cluster_op = get_one_body_cluster_coef(
+        one_body_params, n_orb, restricted=restricted
+    )
+    tril = np.tril_indices(n_orb, k=-1)
+    if restricted:
+        one_body_ham = fqe.get_restricted_hamiltonian((-1j * one_body_cluster_op,))
+    else:
+        one_body_ham = fqe.get_sso_hamiltonian((-1j * one_body_cluster_op,))
+    # Apply U1b^{dag}
+    phi.time_evolve(1, one_body_ham, inplace=True)
+    lam.time_evolve(1, one_body_ham, inplace=True)
+    one_body_grad = np.zeros_like(one_body_params)
+    n_one_body_params = len(one_body_params)
+    grad_position = n_one_body_params - 1
+    for iparam in range(len(one_body_params)):
```

Comment:
Ideally, if we can get line comments for this function, that would be nice.  It looks pretty in depth.